kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -414,260 +415,151 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
     return new Tuple2<>(Math.max(minReaderVersion, 1), Math.max(minWriterVersion, 1));
   }
 
-  /////////////////////////////////////////////////////////////////////////////////
-  /// Everything below will be removed once the Kernel upgrades to use the      ///
-  /// above interfaces.                                                         ///
-  /////////////////////////////////////////////////////////////////////////////////
-  private static final Set<String> SUPPORTED_WRITER_FEATURES =
-      Collections.unmodifiableSet(
-          new HashSet<String>() {
-            {
-              add("appendOnly");
-              add("inCommitTimestamp");
-              add("columnMapping");
-              add("typeWidening-preview");
-              add("typeWidening");
-              add(DOMAIN_METADATA_FEATURE_NAME);
-              add(ROW_TRACKING_FEATURE_NAME);
-            }
-          });
+  /**
+   * Upgrade the current protocol to satisfy all auto-update capable features required by the given
+   * metadata. If the current protocol already satisfies the metadata requirements, return empty.
+   *
+   * @param newMetadata the new metadata to be applied to the table.
+   * @param needDomainMetadataSupport whether the table needs to explicitly support domain metadata.
+   * @param currentProtocol the current protocol of the table.
+   * @return the upgraded protocol and the set of new features that were enabled in the upgrade.
+   */
+  public static Optional<Tuple2<Protocol, Set<TableFeature>>> autoUpgradeProtocolBasedOnMetadata(
+      Metadata newMetadata, boolean needDomainMetadataSupport, Protocol currentProtocol) {
 
-  private static final Set<String> SUPPORTED_READER_FEATURES =
-      Collections.unmodifiableSet(
-          new HashSet<String>() {
-            {
-              add("columnMapping");
-              add("deletionVectors");
-              add("timestampNtz");
-              add("typeWidening-preview");
-              add("typeWidening");
-              add("vacuumProtocolCheck");
-              add("variantType");
-              add("variantType-preview");
-              add("v2Checkpoint");
-            }
-          });
+    Set<TableFeature> autoEnabledFeatures =
```

Comment:
nit: we should decide if this is only `newAutoEnabledFeatures` (in which case we should only add domain_metadata below if it's not already there) or all `autoEnabledFeatures` and update the name as well

I think this is fine to be all autoEnabledFeatures right? maybe update `extractAutomaticallyEnabledNewFeatures` to not remove the existing features

Reply:
makes more semantic sense - create the protocol we require for this table